sys/net/network_layer/ng_ipv6/ng_ipv6.c

```diff
@@ -0,0 +1,501 @@
+
+    if (ng_ipv6_addr_is_multicast(&hdr->dst)) {
+        _ipv6_send_multicast(iface, pkt, ipv6, payload, prep_hdr);
+    }
+    else {
+        ng_ipv6_addr_t *next_hop = NULL;
+
+        next_hop = &hdr->dst;   /* TODO: next hop determination */
+
+        if ((nc_entry = ng_ipv6_nc_get_reachable(iface, next_hop)) == NULL) {
+            DEBUG("ipv6: No link layer address for next_hop %s found.\n",
+                  ng_ipv6_addr_to_str(addr_str, next_hop, sizeof(addr_str)));
+            ng_pktbuf_release(pkt);
+            return;
+        }
+        else {
+            iface = nc_entry->iface;
+        }
+
+        if (iface == KERNEL_PID_UNDEF) {
+            DEBUG("ipv6: no interface for %s registered, dropping packet\n",
+                  ng_ipv6_addr_to_str(addr_str, next_hop, sizeof(addr_str)));
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        if (_fill_ipv6_hdr(iface, ipv6, payload) < 0) {
+            /* error on filling up header */
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        _ipv6_send_unicast(iface, nc_entry->l2_addr, nc_entry->l2_addr_len, pkt);
+    }
+}
+
+/* functions for receiving */
+static inline bool _pkt_not_for_me(kernel_pid_t *iface, ng_ipv6_hdr_t *hdr)
+{
+    if (*iface == KERNEL_PID_UNDEF) {
+        *iface = ng_ipv6_netif_find_by_addr(NULL, &hdr->dst);
+        return (*iface == KERNEL_PID_UNDEF);
+    }
+    else {
+        return (ng_ipv6_netif_find_addr(*iface, &hdr->dst) == NULL);
+    }
+}
+
+static void _dispatch_ipv6_pkt(ng_nettype_t type, uint32_t demux_ctx,
+                               ng_pktsnip_t *pkt)
+{
+    msg_t msg;
+    ng_netreg_entry_t *entry = ng_netreg_lookup(type, demux_ctx);
+
+    msg.type = NG_NETAPI_MSG_TYPE_RCV;
+    msg.content.ptr = (char *)pkt;
+
+    while (entry) {
+        DEBUG("ipv6: Send receive command for %p to %" PRIu16 "\n", (void *)pkt,
+              entry->pid);
+        msg_send(&msg, entry->pid);
+        entry = ng_netreg_getnext(entry);
+    }
+}
+
+static void _ipv6_receive(kernel_pid_t iface, ng_pktsnip_t *pkt)
+{
+    ng_pktsnip_t *ipv6, *payload;
+    ng_ipv6_hdr_t *hdr;
+
+    LL_SEARCH_SCALAR(pkt, payload, type, NG_NETTYPE_IPV6);
+
+    /* is it a valid IPv6 packet? */
+
+    if (payload == NULL) {
+        DEBUG("ipv6: unable to get write access to packet, dropping packet\n");
+        return;
+    }
+
+    if ((payload->next != NULL) && (payload->next->type == NG_NETTYPE_IPV6) &&
+        (payload->next->size == sizeof(ng_ipv6_hdr_t))) {
+        /* IP header was already marked. Take it. */
+        ipv6 = payload->next->data;
+
+        if (!ng_ipv6_hdr_is_ipv6_hdr(ipv6->data)) {
+            DEBUG("ipv6: Received packet was not IPv6, dropping packet\n");
+            ng_pktbuf_release(pkt);
+            return;
+        }
+
+        payload = _start_write(payload);
+    }
+    else {
+        if (!ng_ipv6_hdr_is_ipv6_hdr(payload->data)) {
```

Comment:
I know. That's why I called 6LoWPAN check functions e.g. `ng_sixlowpan_frag_is()`. I can do this for this function too, but I would prefer to do this as a follow-up.